[quantization] Use torchbind for Linear and RNN PackedParams

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34140 [quantization] Use torchbind for Linear and RNN PackedParams**

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

Commit:
[WIP][quantization] Use torchbind for Linear PackedParams

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"




[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -197,3 +205,75 @@ Tensor ConvertToChannelsLast3dTensor(const Tensor& src) {
+                return std::make_tuple(
+                    std::move(weight),
+                    std::move(bias),
+                    params->backend(),
```
same comment as before: we should NOT serialize the backend and instead infer it from qengine setting at the time of loading

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -197,3 +205,75 @@ Tensor ConvertToChannelsLast3dTensor(const Tensor& src) {
+                bit_width = std::move(std::get<3>(state));
+
+#ifdef USE_FBGEMM
+                if (backend == "FBGEMM") {
```
just check qengine setting here

dzhulgakov commented on torch/csrc/jit/passes/onnx/unpack_quantized_weights.cpp:
```diff
@@ -145,13 +145,24 @@ void unpackQuantizedWeightsHelper(
+    c10::optional<at::Tensor> bias;
+
+    if (itr->second.isTuple()) {
+      // Pre-unpacked weights. Comes from Linear weights which are
```
this is a magical dependency on the LinearWeights format. At least put a comment there that if it changes - you need to update this place too (or make it a static method of LinearWeight class)

dzhulgakov replied:
so why you can't pass the custom class all the way here and call `__getstate__` here? There's a nice symmetry then with the 'else' case which also does unpacking (this is pretty much the same as the comment above)

jamesr66a replied:
It doesn't fit into the existing infrastructure:
```
======================================================================
ERROR: test_torchbind_lower_graph (__main__.TestScript)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_jit.py", line 4630, in test_torchbind_lower_graph
    onnx_model = export_to_onnx(model, x, input_names)
  File "test/test_jit.py", line 4624, in export_to_onnx
    operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
  File "/data/users/jamesreed/pytorch/torch/onnx/__init__.py", line 172, in export
    custom_opsets, enable_onnx_checker, use_external_data_format)
  File "/data/users/jamesreed/pytorch/torch/onnx/utils.py", line 92, in export
    use_external_data_format=use_external_data_format)
  File "/data/users/jamesreed/pytorch/torch/onnx/utils.py", line 519, in _export
    fixed_batch_size=fixed_batch_size)
  File "/data/users/jamesreed/pytorch/torch/onnx/utils.py", line 340, in _model_to_graph
    in_vars, in_desc = torch.jit._flatten(tuple(args) + tuple(params))
RuntimeError: Only tuples, lists and Variables supported as JIT inputs/outputs. Dictionaries and strings are also accepted but their usage is not recommended. But got unsupported type torch._C.ScriptObject
```

And I don't want to rabbit-hole down another fractal of redesign when this solution just works

dzhulgakov commented on torch/csrc/jit/passes/lower_graph.cpp:
```diff
@@ -114,16 +116,30 @@ std::pair<std::shared_ptr<Graph>, std::vector<Slot>> lower_graph(
+          type->python_str(),
+          " encountered in graph lowering. This type is not supported in ONNX export.");
+      result.emplace_back(
+          script::Object(obj.toObject()).run_method("__getstate__"));
```
just curious - why do we need to call _getstate_ here instead of keeping the class as is and invoking `__getstate__` in unpackQuantizedWeightsHelper (so that there's one less magic place)

dzhulgakov replied:
question still stands

dzhulgakov commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -400,19 +400,44 @@ void initJITBindings(PyObject* module) {
+            return retval;
+          })
+      .def("_jit_decay_packed_param_input_types", [](Graph& g) {
+        for (Value* i : g.inputs()) {
```
why is it a separate pass? can't `unpackQuantizedWeightsHelper ` handle it? it logically belongs there (when we replace class instance with tensor)

jamesr66a replied:
I don't really follow where you're referring to when you say `unpackQuantizedWeightsHelper` should handle it

dzhulgakov replied:
I meant that unpackQuantizedWeightsHelper does the modification of params. So it's logical for it to also set the graph input to be of the correct type (or maybe delete the input all together). Otherwise these two passes (_jit_pass_filter_non_tensor_arguments and _jit_decay_packed_param_input_types) are a bit magical to be called from the top level

jamesr66a replied:
It doesn't fit into the existing infrastructure:
```
======================================================================
ERROR: test_torchbind_lower_graph (__main__.TestScript)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_jit.py", line 4630, in test_torchbind_lower_graph
    onnx_model = export_to_onnx(model, x, input_names)
  File "test/test_jit.py", line 4624, in export_to_onnx
    operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
  File "/data/users/jamesreed/pytorch/torch/onnx/__init__.py", line 172, in export
    custom_opsets, enable_onnx_checker, use_external_data_format)
  File "/data/users/jamesreed/pytorch/torch/onnx/utils.py", line 92, in export
    use_external_data_format=use_external_data_format)
  File "/data/users/jamesreed/pytorch/torch/onnx/utils.py", line 519, in _export
    fixed_batch_size=fixed_batch_size)
  File "/data/users/jamesreed/pytorch/torch/onnx/utils.py", line 340, in _model_to_graph
    in_vars, in_desc = torch.jit._flatten(tuple(args) + tuple(params))
RuntimeError: Only tuples, lists and Variables supported as JIT inputs/outputs. Dictionaries and strings are also accepted but their usage is not recommended. But got unsupported type torch._C.ScriptObject
```

And I don't want to rabbit-hole down another fractal of redesign when this solution just works

dzhulgakov commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -476,6 +476,14 @@ static void setInputTensorTypes(Graph& g, const Stack& stack, bool complete) {
   auto s_iter = stack.begin();
   for (auto v : input_values) {
     AT_ASSERT(s_iter != stack.end());
+    // Leave packed param types alone. This is needed for downstream passes
```
shall we just handle all custom classes like that here? trying to reduce amount of hard-coding

dzhulgakov commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -259,6 +261,37 @@ struct SourceImporterImpl : public Resolver,
+            "__torch__.torch.nn.quantized.modules.linear.Linear") {
+      auto lhs = Var(assign.lhs());
+      if (!assign.type().present() || assign.type().get().kind() != TK_VAR) {
+        ;
```
? :)

dzhulgakov commented on test/test_jit.py:
```diff
@@ -5293,6 +5293,77 @@ def test_torchbind_pickle_serialization(self):
+            model = torch.jit.load(buf)
+            f = io.BytesIO()
+            torch.onnx.export(model, input, f, input_names=input_names, example_outputs=outputs,
+                              operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```
you probably don't need operator_export_type here, do you?

jamesr66a replied:
This is what I get:

```
RuntimeError: Exporting the operator quantize_per_tensor to ONNX opset version 9 is not supported. Please open a bug to request ONNX export support for the missing operator.

```

dzhulgakov reviewed (commented):
Wow, collection of patches to make onnx work is magical - great work! I added a few questions on how maybe simplify special-casing, but not sure all of them are valid.

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

kimishpatel commented:
You may see mobile build breaking with custom class. I am landing this PR, https://github.com/pytorch/pytorch/pull/34368, to fix that. Just FYI.

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

dzhulgakov reviewed (changes requested):
please address comments on not serializing 'backend'

jamesr66a commented:
@dzhulgakov this is still WIP and I didn't re-request review

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

jerryzh168 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -42,26 +42,45 @@ def _weight_bias(self):
-        destination[prefix + 'weight'] = w
-        destination[prefix + 'bias'] = b
         destination[prefix + 'dtype'] = self.dtype
+        destination[prefix + '_packed_params'] = self._weight_bias()
```
should this be `destination[prefix + '_packed_params'] = self._packed_params()`?

jamesr66a replied:
No, this is handling unpacking the weights for serialization

jerryzh168 replied:
why do you unpack here? I think you can do this unpacking even without the torchbind class support right? I thought we want to save the packed param directly here

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

jerryzh168 commented on test/test_jit.py:
```diff
@@ -5938,6 +5938,83 @@ def forward(self) -> str:
+            def __init__(self):
+                super(ConvModel, self).__init__()
+                self.qconfig = torch.quantization.default_qconfig
+                self.fc1 = torch.quantization.QuantWrapper(torch.nn.Conv2d(3, 5, 2, bias=True).to(dtype=torch.float))
```
nit: fc1 -> conv

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

jamesr66a commented on torch/_overrides.py:
```diff
@@ -546,11 +546,11 @@ def get_testing_overrides():
-        torch.quantized_lstm: (lambda input, hx, params, has_biases, num_layers, dropout, train, bidirectional,
-                               batch_first, dtype=None, use_dynamic=False: -1),
+        # torch.quantized_lstm: (lambda input, hx, params, has_biases, num_layers, dropout, train, bidirectional,
+        #                       batch_first, dtype=None, use_dynamic=False: -1),
```
@ngoldbaum @ezyang These functions no longer take `Tensor`s for their parameter arguments, which was now tripping up the `test_override` tests as it could not cast `TensorLike` to the appropriate TorchBind object. I removed them from testing for now. Let me know if there's something else we should do here

ezyang replied:
cc @hameerabbasi

hameerabbasi replied:
If they no longer take `Tensor`s, they're no longer supported by `__torch_function__`. This is a backwards incompatible change as these functions can no longer be overridden. May I ask what the motivation was here?

jamesr66a replied:
These were never tensors, they were typed as Tensor but what it actually was is a C++ struct punned into 1-d ByteTensors. This is making that data type explicit now that we have more rich data types in our type system

ezyang replied:
OK, sounds to me that dropping them from override test is the right thing to do then!

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

ezyang commented:
FYI if I ever get to land the torch/library.h rename, these includes need to get updated to torch/library.h

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Figure out BC story
2) Quantized RNN test is flaky

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Quantized RNN test is flaky. cc @raghuramank100 

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Quantized RNN test is flaky. cc @raghuramank100 

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[WIP][quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Quantized RNN test is flaky. cc @raghuramank100 

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/native/RNN.cpp:
```diff
@@ -654,11 +558,30 @@ static c10::List<c10::intrusive_ptr<CellParamsBase>> gather_quantized_params_fp1
+    //
+    // Hopefully we can remove this in the future when we eliminate
+    // the old style altogether
+    packed_struct_ih->set_bias(params[i + 2]);
```
Hm, is it thread-safe? because technically we can call RNN from multiple threads at the same time and it could race.

In practice though it should be fine as we're just setting the pointer to the same value (so all refcounts will check out)

jamesr66a replied:
What should we do here? lock?

dzhulgakov commented on torch/csrc/jit/passes/lower_graph.cpp:
```diff
@@ -113,16 +115,31 @@ std::pair<std::shared_ptr<Graph>, std::vector<Slot>> lower_graph(
-    result.emplace_back(slot.obj->getSlot(slot.offset).toTensor());
+    auto obj = slot.obj->getSlot(slot.offset);
+    if (obj.isTensor()) {
+      result.emplace_back(obj.toTensor());
```
nit: toTensor is not needed as both are just IValues now

dzhulgakov reviewed (approved):
Ok, ship it if all tests pass.

I still have questions on whether onnx export can be organized a bit better, but it's a non-blocker

ezyang commented on aten/src/ATen/native/quantized/library.cpp:
```diff
@@ -61,8 +65,10 @@ TORCH_LIBRARY(_quantized, m) {
-  m.def("linear_dynamic(Tensor X, Tensor W_prepack) -> Tensor Y");
-  m.def("linear_prepack(Tensor W, Tensor? B=None) -> Tensor W_prepack");
-  m.def("linear_prepack_fp16(Tensor W, Tensor? B=None) -> Tensor W_prepack");
+  m.def("linear(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack, float Y_scale_i, int Y_zero_point_i) -> Tensor Y");
```
I'm surprised this works, because IIRC all types mentioned in schemas must be defined before you try to register the schema. Maybe a latent SIOF failure here. Better to define the class directly in this `TORCH_LIBRARY` block

ezyang commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -2215,10 +2215,39 @@ struct to_ir {
+        // a TorchBind class and not a cpp_custom_type_hack tensor anymore, but the
+        // old serialized models are tightly coupled with the type_hack version.
+        // If we still create a Tensor here, then the quantized_lstm.legacy overload
+        // can kick in in forward_impl(), and the module will still run correctly.
```
Under what situation can this BC hack be removed?

jamesr66a replied:
Either of:
- We don't see any production models with this serialized pattern anymore, or
- We make a call that we're no longer supporting that old version of the quantized RNN module in OSS

ezyang commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -197,3 +205,65 @@ Tensor ConvertToChannelsLast3dTensor(const Tensor& src) {
+torch::jit::class_<LinearPackedParamsBase> register_linear_params() {
+  using SerializationType = std::tuple<at::Tensor, c10::optional<at::Tensor>>;
+  static auto register_linear_params =
+      torch::jit::class_<LinearPackedParamsBase>("quantized", "LinearPackedParamsBase")
```
This can be put into the `TORCH_LIBRARY` block now. Just say:

```
m.class_<LinearPackedParamsBase>("LinearPackedParamsBase")
  ...
```

LMK if dependency ordering means you can't conveniently put this in the other C++ file.

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Quantized RNN test is flaky. cc @raghuramank100 

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Quantized RNN test is flaky. cc @raghuramank100 

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Quantized RNN test is flaky. cc @raghuramank100 

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Quantized RNN test is flaky. cc @raghuramank100 

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Quantized RNN test is flaky. cc @raghuramank100 

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Quantized RNN test is flaky. cc @raghuramank100 

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Quantized RNN test is flaky. cc @raghuramank100 

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Current issues:
1) Quantized RNN test is flaky. cc @raghuramank100 

Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/fbgemm_utils.cpp:
```diff
@@ -3,8 +3,6 @@
 #include <ATen/native/quantized/cpu/conv_packed_params.h>
```
feel free to merge conv_packed_params.h with packed_params.h as well

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

Commit:
Update on "[quantization] Use torchbind for Linear and RNN PackedParams"


Differential Revision: [D20229168](https://our.internmc.facebook.com/intern/diff/D20229168)

[ghstack-poisoned]

malfet commented:
Reverting, as this broke windows builds with ([full log](https://app.circleci.com/pipelines/github/pytorch/pytorch/166664/workflows/efca978c-85c6-4462-bc68-19783d8dcb92/jobs/5398557/steps)):
```
[1482/3335] C:\Users\circleci\project\build\win_tmp\bin\sccache-cl.exe   /TP -DAT_PARALLEL_OPENMP=1 -DCPUINFO_SUPPORTED_PLATFORM=1 -DFMT_HEADER_ONLY=1 -DIDEEP_USE_MKL -DMAGMA_V2 -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MKL -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -Dtorch_cpu_EXPORTS -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -Icaffe2\contrib\aten -I..\third_party\onnx -Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -I..\caffe2\..\torch\csrc\api -I..\caffe2\..\torch\csrc\api\include -I..\caffe2\aten\src\TH -Icaffe2\aten\src\TH -I..\caffe2\..\torch\..\aten\src -Icaffe2\aten\src -Icaffe2\..\aten\src -Icaffe2\..\aten\src\ATen -I..\caffe2\..\torch\csrc -I..\caffe2\..\torch\..\third_party\miniz-2.0.8 -I..\aten\src\TH -I..\aten\..\third_party\catch\single_include -I..\aten\src\ATen\.. -Icaffe2\aten\src\ATen -I..\third_party\miniz-2.0.8 -I..\caffe2\core\nomnigraph\include -I..\c10\.. -Ithird_party\ideep\mkl-dnn\include -I..\third_party\ideep\mkl-dnn\src\..\include -I..\third_party\cpuinfo\include -I..\third_party\FP16\include -I..\third_party\fmt\include -I..\cmake\..\third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I..\third_party\protobuf\src -Iwin_tmp\mkl\include -I..\cmake\..\third_party\eigen -IC:\Jenkins\Miniconda3\include -IC:\Jenkins\Miniconda3\lib\site-packages\numpy\core\include -I..\cmake\..\third_party\pybind11\include -I..\cmake\..\third_party\cub -Iwin_tmp\magma\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" -I..\third_party\ideep\mkl-dnn\include -I..\third_party\ideep\include -I..\caffe2 /DWIN32 /D_WINDOWS /GR /EHsc /w /bigobj -openmp -DNDEBUG -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2 /DNDEBUG /w /bigobj -DNDEBUG   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHsc /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\aten\src\ATen\native\RNN.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\aten\src\ATen\native\RNN.cpp
FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.obj 
C:\Users\circleci\project\build\win_tmp\bin\sccache-cl.exe   /TP -DAT_PARALLEL_OPENMP=1 -DCPUINFO_SUPPORTED_PLATFORM=1 -DFMT_HEADER_ONLY=1 -DIDEEP_USE_MKL -DMAGMA_V2 -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MKL -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -Dtorch_cpu_EXPORTS -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -Icaffe2\contrib\aten -I..\third_party\onnx -Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -I..\caffe2\..\torch\csrc\api -I..\caffe2\..\torch\csrc\api\include -I..\caffe2\aten\src\TH -Icaffe2\aten\src\TH -I..\caffe2\..\torch\..\aten\src -Icaffe2\aten\src -Icaffe2\..\aten\src -Icaffe2\..\aten\src\ATen -I..\caffe2\..\torch\csrc -I..\caffe2\..\torch\..\third_party\miniz-2.0.8 -I..\aten\src\TH -I..\aten\..\third_party\catch\single_include -I..\aten\src\ATen\.. -Icaffe2\aten\src\ATen -I..\third_party\miniz-2.0.8 -I..\caffe2\core\nomnigraph\include -I..\c10\.. -Ithird_party\ideep\mkl-dnn\include -I..\third_party\ideep\mkl-dnn\src\..\include -I..\third_party\cpuinfo\include -I..\third_party\FP16\include -I..\third_party\fmt\include -I..\cmake\..\third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I..\third_party\protobuf\src -Iwin_tmp\mkl\include -I..\cmake\..\third_party\eigen -IC:\Jenkins\Miniconda3\include -IC:\Jenkins\Miniconda3\lib\site-packages\numpy\core\include -I..\cmake\..\third_party\pybind11\include -I..\cmake\..\third_party\cub -Iwin_tmp\magma\include -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include" -I..\third_party\ideep\mkl-dnn\include -I..\third_party\ideep\include -I..\caffe2 /DWIN32 /D_WINDOWS /GR /EHsc /w /bigobj -openmp -DNDEBUG -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2 /DNDEBUG /w /bigobj -DNDEBUG   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHsc /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\aten\src\ATen\native\RNN.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\aten\src\ATen\native\RNN.cpp
..\aten\src\ATen\native\RNN.cpp(12): error C2065: 'LinearPackedParamsBase': undeclared identifier
..\aten\src\ATen\native\RNN.cpp(12): error C2923: 'torch::class_': 'LinearPackedParamsBase' is not a valid template type argument for parameter 'CurClass'
..\aten\src\ATen\native\RNN.cpp(64): error C2065: 'LinearPackedParamsBase': undeclared identifier
..\aten\src\ATen\native\RNN.cpp(64): error C2923: 'c10::intrusive_ptr': 'LinearPackedParamsBase' is not a valid template type argument for parameter 'TTarget'
..\aten\src\ATen\native\RNN.cpp(64): error C3203: 'intrusive_target_default_null_type': unspecialized class template can't be used as a template argument for template parameter 'NullType', expected a real type
..\aten\src\ATen\native\RNN.cpp(64): error C3203: 'intrusive_ptr': unspecialized class template can't be used as a template argument for template parameter '_Ty', expected a real type
..\aten\src\ATen\native\RNN.cpp(286): error C2065: 'LinearPackedParamsBase': undeclared identifier
..\aten\src\ATen\native\RNN.cpp(286): error C2923: 'c10::intrusive_ptr': 'LinearPackedParamsBase' is not a valid template type argument for parameter 'TTarget'
..\aten\src\ATen\native\RNN.cpp(286): error C3203: 'intrusive_target_default_null_type': unspecialized class template can't be used as a template argument for template parameter 'NullType', expected a real type
..\aten\src\ATen\native\RNN.cpp(287): error C2065: 'LinearPackedParamsBase': undeclared identifier
..\aten\src\ATen\native\RNN.cpp(287): error C2923: 'c10::intrusive_ptr': 'LinearPackedParamsBase' is not a valid template type argument for parameter 'TTarget'
..\aten\src\ATen\native\RNN.cpp(287): error C3203: 'intrusive_target_default_null_type': unspecialized class template can't be used as a template argument for template parameter 'NullType', expected a real type
..\aten\src\ATen\native\RNN.cpp(293): error C2065: 'LinearPackedParamsBase': undeclared identifier
..\aten\src\ATen\native\RNN.cpp(292): error C2923: 'c10::intrusive_ptr': 'LinearPackedParamsBase' is not a valid template type argument for parameter 'TTarget'
..\aten\src\ATen\native\RNN.cpp(292): error C3203: 'intrusive_target_default_null_type': unspecialized class template can't be used as a template argument for template parameter 'NullType', expected a real type
..\aten\src\ATen\native\RNN.cpp(295): error C2065: 'LinearPackedParamsBase': undeclared identifier
..\aten\src\ATen\native\RNN.cpp(304): error C2065: 'LinearPackedParamsBase': undeclared identifier
..\aten\src\ATen\native\RNN.cpp(304): error C2923: 'c10::intrusive_ptr': 'LinearPackedParamsBase' is not a valid template type argument for parameter 'TTarget'
..\aten\src\ATen\native\RNN.cpp(304): error C3203: 'intrusive_target_default_null_type': unspecialized class template can't be used as a template argument for template parameter 'NullType', expected a real type
..\aten\src\ATen\native\RNN.cpp(305): error C2065: 'LinearPackedParamsBase': undeclared identifier
..\aten\src\ATen\native\RNN.cpp(305): error C2923: 'c10::intrusive_ptr': 'LinearPackedParamsBase' is not a valid template type argument for parameter 'TTarget'
..\aten\src\ATen\native\RNN.cpp(305): error C3203: 'intrusive_target_default_null_type': unspecialized class template can't be used as a template argument for template parameter 'NullType', expected a real type
..\aten\src\ATen\native\RNN.cpp(292): error C2672: 'std::move': no matching overloaded function found
..\aten\src\ATen\native\RNN.cpp(292): error C2893: Failed to specialize function template 'remove_reference<_Ty>::type &&std::move(_Ty &&) noexcept'
..\aten\src\ATen\native\RNN.cpp(292): note: With the following template arguments:
..\aten\src\ATen\native\RNN.cpp(292): note: '_Ty=c10::intrusive_ptr &'
..\aten\src\ATen\native\RNN.cpp(300): error C2672: 'std::move': no matching overloaded function found
..\aten\src\ATen\native\RNN.cpp(300): error C2893: Failed to specialize function template 'remove_reference<_Ty>::type &&std::move(_Ty &&) noexcept'
..\aten\src\ATen\native\RNN.cpp(300): note: With the following template arguments:
..\aten\src\ATen\native\RNN.cpp(300): note: '_Ty=c10::intrusive_ptr &'
..\aten\src\ATen\native\RNN.cpp(302): fatal error C1903: unable to recover from previous error(s); stopping compilation
```
